### PR TITLE
[CAMEL-20728] add S3 multipart upload support for stream producer in camel-aws2-s3

### PR DESCRIPTION
**Pull request by bonnetb**
# Description

See: https://issues.apache.org/jira/browse/CAMEL-20728
This PR aims to provide camel-aws2s3 producer for support for multipart upload in combination with streaming mode. Actually, today's streaming mode agregates messages into S3 objects, but each object is sent as a whole, which can be an issue if it is huge (memory failure). This PR enables multipart (S3 API multipart) for each object upload.
multiPartUpload must be enabled, and partSize will specify chunk sizes.

### CODE REVIEW

**Review by oscerd — APPROVED** [review]

**Comment by oscerd** [comment]
@jono-morris can you have a look? You worked on the streaming upload s3 for a while and also in relation to multipart. Thanks.

File: components/camel-aws/camel-aws2-s3/src/main/java/org/apache/camel/component/aws2/s3/stream/AWS2S3StreamUploadProducer.java [hunk 1]
```diff
@@ -244,29 +262,41 @@ private CompleteMultipartUploadResponse completeUpload(UploadState state) {
...
+        try {
+            final CompleteMultipartUploadResponse uploadResult
+                    = getEndpoint().getS3Client().completeMultipartUpload(compRequest);
 
-        // Converting the index to String can cause extra overhead
-        if (LOG.isInfoEnabled()) {
-            LOG.info("Completed upload for the part {} with etag {} at index {}", part, uploadResult.eTag(),
-                    state.index);
+            // Converting the index to String can cause extra overhead
+            if (LOG.isInfoEnabled()) {
```
Comment (davsclaus) [inline components/camel-aws/camel-aws2-s3/src/main/java/org/apache/camel/component/aws2/s3/stream/AWS2S3StreamUploadProducer.java:270]: Change to DEBUG logging

File: components/camel-aws/camel-aws2-s3/src/main/java/org/apache/camel/component/aws2/s3/stream/AWS2S3StreamUploadProducer.java [hunk 2]
```diff
@@ -244,29 +262,41 @@ private CompleteMultipartUploadResponse completeUpload(UploadState state) {
...
+                LOG.info("Completed upload for the part {}, multipart {} with etag {} at index {}", part, state.multipartIndex,
+                        uploadResult.eTag(),
+                        state.index);
+            }
+            part.getAndIncrement();
+            return uploadResult;
+        } catch (Exception e) {
+            LOG.error("Error completing multipart updload");
```
Comment (davsclaus) [inline components/camel-aws/camel-aws2-s3/src/main/java/org/apache/camel/component/aws2/s3/stream/AWS2S3StreamUploadProducer.java:278]: Change to WARN and we likely need to include the caused exception so the user has a chance to understand why it failed. Also adjust the message to say that the upload will be aborted.

**Review by davsclaus — CHANGES_REQUESTED** [review]

**Review by davsclaus — APPROVED** [review]

**Pull request merged**